src/v/pandaproxy/client/consumer.h

```diff
@@ -0,0 +1,558 @@
+              [](model::partition_id id, kafka::response_writer& writer) {
+                  writer.write(id);
+              });
+        }
+
+        return kafka::sync_group_request_assignment{
+          .member_id = m.first, .assignment = iobuf_to_bytes(assignments_buf)};
+    };
+
+    std::vector<kafka::sync_group_request_assignment>
+    encode(const assignments& assignments) const {
+        std::vector<kafka::sync_group_request_assignment> result;
+        result.reserve(assignments.size());
+        for (const auto& m : assignments) {
+            result.push_back(encode(m));
+        }
+        return result;
+    }
+
+    assignment decode(const bytes& b) const {
+        if (b.empty()) {
+            return {};
+        }
+        kafka::request_reader reader(bytes_to_iobuf(b));
+        auto result = reader.read_array([](kafka::request_reader& reader) {
+            auto topic = model::topic(reader.read_string());
+            return std::make_pair(
+              std::move(topic),
+              reader.read_array([](kafka::request_reader& reader) {
+                  return model::partition_id(reader.read_int32());
+              }));
+        });
+        return assignment(
+          std::make_move_iterator(result.begin()),
+          std::make_move_iterator(result.end()));
+    }
+};
+
+struct assignment_range : public assignment_plan {
+    assignments plan(
+      const std::vector<kafka::member_id>& members,
+      const std::vector<kafka::metadata_response::topic>& topics) final {
+        assignments assignments;
+        for (auto const& t : topics) {
+            auto [len, rem] = std::ldiv(t.partitions.size(), members.size());
+            auto iterations = std::min(t.partitions.size(), members.size());
+            auto p_begin = t.partitions.begin();
+            auto p_end = p_begin;
+            auto mem_it = members.begin();
+            while (iterations--) {
+                p_end += (rem > 0) ? (len + !!(rem--)) : len;
+                auto& rtm = assignments[*mem_it][t.name];
+                rtm.reserve(std::distance(p_begin, p_end));
+                std::transform(
+                  p_begin, p_end, std::back_inserter(rtm), [](auto& p) {
+                      return p.index;
+                  });
+                p_begin = p_end;
+                ++mem_it;
+            }
+        }
+        return assignments;
+    }
+};
+
+inline auto make_find_coordinator_request(kafka::group_id group_id) {
+    return [group_id{std::move(group_id)}]() {
+        return kafka::find_coordinator_request(group_id);
+    };
+}
+
+inline auto make_list_groups_request() {
+    return []() { return kafka::list_groups_request{}; };
+}
+
+inline auto make_describe_group_request(kafka::group_id group_id) {
+    return [group_id{std::move(group_id)}]() {
+        kafka::describe_groups_request req;
+        req.data.groups.push_back(group_id);
+        return req;
+    };
+}
+
+inline kafka::join_group_request_protocol
```

Comment:
is this in the hot path? all of these inline methods  seem hard to understand what *is* the actual hot path that we want to optimize vs putting it on the  .cc file.